### DISCUSSION TITLE
`Unable to validate source` on `mc cp`

### DISCUSSION THREAD

**Discussion by qyz** [minio/minio#14011] [Q&A] [answered]
Hi,

I'm doing backups with restic to a bare-metal machine running MinIO. Today restic was unable to retrieve an older object.

I tried to narrow this down. No hardware errors were logged, S.M.A.R.T. data and underlying filesystems check out ok. The objects, which restic complains about, also show unexpected behaviour when I try to retrieve them with mc:

## Expected Behavior
I can retrieve an object with `mc cp`.

## Current Behavior
`mc cp` cannot retrieve the object, although `mc ls` says it exists. 

```
root@phoenix:~# mc ls phoenix/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194
[2020-06-13 18:09:54 CEST] 4.8MiB 7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194

root@phoenix:~# mc cp phoenix/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194 /tmp
mc: <ERROR> Unable to validate source `phoenix/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194`.
```

## Possible Solution

I wonder whether this behaviour has something to do with issue #13397. I had similar problems back then, which were addressed in excellence by @harshavardhana.

The object in question here seems to use different storage formats on the different disks. Please note the timestamps. `Jun 13  2020` seems to be the creation time of the object. At `Dec 27 22:03` restic requested this object.

```
root@phoenix:/minio_a# ls -l $(find disk?/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194 -type f)
-rw-r--r-- 1 minio-user minio-user 831467 Jun 13  2020 disk1/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194/part.1
-rw-r--r-- 1 minio-user minio-user    510 Jun 13  2020 disk1/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194/xl.json
-rw-r--r-- 1 minio-user minio-user 831467 Dec 27 22:03 disk2/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194/84e24913-396a-4bb7-a727
-59ec817fdf00/part.1
-rw-r--r-- 1 minio-user minio-user    364 Dec 27 22:03 disk2/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194/xl.meta
-rw-r--r-- 1 minio-user minio-user 831467 Dec 27 22:03 disk3/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194/84e24913-396a-4bb7-a727
-59ec817fdf00/part.1
-rw-r--r-- 1 minio-user minio-user    364 Dec 27 22:03 disk3/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194/xl.meta
-rw-r--r-- 1 minio-user minio-user 831467 Dec 27 22:03 disk4/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194/84e24913-396a-4bb7-a727
-59ec817fdf00/part.1
-rw-r--r-- 1 minio-user minio-user    364 Dec 27 22:03 disk4/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194/xl.meta
-rw-r--r-- 1 minio-user minio-user 831467 Dec 27 22:03 disk5/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194/84e24913-396a-4bb7-a727
-59ec817fdf00/part.1
-rw-r--r-- 1 minio-user minio-user    364 Dec 27 22:03 disk5/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194/xl.meta
-rw-r--r-- 1 minio-user minio-user 831467 Dec 27 22:03 disk6/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194/84e24913-396a-4bb7-a727
-59ec817fdf00/part.1
-rw-r--r-- 1 minio-user minio-user    364 Dec 27 22:03 disk6/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194/xl.meta
-rw-r--r-- 1 minio-user minio-user 831467 Jun 13  2020 disk7/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194/part.1
-rw-r--r-- 1 minio-user minio-user    510 Jun 13  2020 disk7/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194/xl.json
-rw-r--r-- 1 minio-user minio-user 831467 Jun 13  2020 disk8/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194/part.1
-rw-r--r-- 1 minio-user minio-user    510 Jun 13  2020 disk8/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194/xl.json
-rw-r--r-- 1 minio-user minio-user 831467 Jun 13  2020 disk9/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194/part.1
-rw-r--r-- 1 minio-user minio-user    510 Jun 13  2020 disk9/a/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194/xl.json
```

## More Debug

```
root@phoenix:~# mc admin trace -v phoenix 2>&1 | tee mc_admin_trace_v.log &
root@phoenix:~# mc cp --debug phoenix/james-full/data/7b/7ba9cb023318883ffb59704974dad6b9b7bc7274f34a4b9569f4b859b38de194 /tmp 2>&1 | tee mc_cp_debug.log
```

[mc_admin_trace_v.log](https://github.com/minio/minio/files/7790575/mc_admin_trace_v.log)
[mc_cp_debug.log](https://github.com/minio/minio/files/7790576/mc_cp_debug.log)

## My Environment
* minio version is `RELEASE.2021-12-20T22-07-16Z`
* Debian 11 on bare metal, 9x2TB disks, ext4, default kernel, AMD Opteron 2378, 32GB ECC RAM.
* System is used for cold, last-resort backups, it stays completely disconnected and I only power it on about once a month for a few hours.

I`d appreciate any help,
Cheers,
qyz

**Comment by harshavardhana** [comment]
Looks like your have objects from legacy format that were not migrated successfully. 

I don't think they are recoverable anymore.

**Comment by harshavardhana** [comment, marked as answer]
> I wonder whether this behaviour has something to do with issue #13397. I had similar problems back then, which were addressed in excellence by @harshavardhana.

These are not related to that issue @qyz - the issue is that your drives were not present or mounted. There seems to be some kind of migration that happened and for some reason there was a crash (or server failure) that failed the migration of all the erasure shards.

We have upgrade tests to ensure that this doesn't happen https://github.com/minio/minio/blob/master/.github/workflows/upgrade-ci-cd.yaml 

But unfortunately in your case, it would need manual work to move things around. Looks like you did something recently that caused it.
